File: matiasmasso/Mat.NET/DT2/DTOBalance.vb
Public Class DTOBalance

    Property emp As DTOEmp
    Property yearMonthFrom As DTOYearMonth
    Property yearMonthTo As DTOYearMonth
    Property items As List(Of DTOPgcClass)

    Shared Function Factory(oEmp As DTOEmp, YearMonthFrom As DTOYearMonth, YearMonthTo As DTOYearMonth) As DTOBalance
        Dim retval As New DTOBalance
        With retval
            .emp = oEmp
            .yearMonthFrom = YearMonthFrom
            .yearMonthTo = YearMonthTo
            .items = New List(Of DTOPgcClass)
        End With
        Return retval
    End Function

    Public Function YearMonths(oPgcClass As DTOPgcClass) As List(Of DTOYearMonth)
        Dim oCtas As New List(Of DTOPgcCta)
        If oPgcClass.Ctas IsNot Nothing AndAlso oPgcClass.Ctas.Count > 0 Then
            oCtas = oPgcClass.Ctas
        Else
            GetPgcCtas(oPgcClass, oCtas)
        End If
        Dim oYearMonths = oCtas.SelectMany(Function(x) x.yearMonths).ToList
        Dim retval = oYearMonths.GroupBy(Function(g) New With {Key g.year, Key g.month}).Select(Function(group) New DTOYearMonth With {.year = group.Key.year, .month = group.Key.month, .Eur = group.Sum(Function(x) x.Eur)}).ToList
        Return retval
    End Function

    Public Function YearMonths(oCod As DTOPgcClass.Cods) As List(Of DTOYearMonth)
        Dim oCtas = Ctas(oCod)
        Dim oYearMonths = oCtas.SelectMany(Function(x) x.yearMonths).ToList
        Dim retval = oYearMonths.GroupBy(Function(g) New With {Key g.year, Key g.month}).Select(Function(group) New DTOYearMonth With {.year = group.Key.year, .month = group.Key.month, .Eur = group.Sum(Function(x) x.Eur)}).ToList
        Return retval
    End Function

    Public Function Ctas(oCod As DTOPgcClass.Cods) As List(Of DTOPgcCta)
        Dim retval As New List(Of DTOPgcCta)
        Dim oClass = GetPgcClass(oCod)
        If oClass IsNot Nothing Then
            GetPgcCtas(oClass, retval)
        End If
        Return retval
    End Function

    Public Function GetPgcClass(oCod As DTOPgcClass.Cods, Optional oParentsFrom As List(Of DTOPgcClass) = Nothing) As DTOPgcClass
        Dim retval As DTOPgcClass = Nothing

        If oParentsFrom Is Nothing Then
            oParentsFrom = _items
        End If

        For Each oClass As DTOPgcClass In oParentsFrom
            If oClass.Cod = oCod Then
                retval = oClass
                Exit For
            ElseIf oClass.Children.Count > 0 Then
                retval = GetPgcClass(oCod, oClass.Children)
                If retval IsNot Nothing Then Exit For
            End If
        Next
        Return retval
    End Function

    Public Sub GetPgcCtas(oParentFrom As DTOPgcClass, oCtas As List(Of DTOPgcCta))
        For Each oClass As DTOPgcClass In oParentFrom.Children
            If oClass.Ctas IsNot Nothing AndAlso oClass.Ctas.Count > 0 Then
                oCtas.AddRange(oClass.Ctas)
            End If
            If oClass.Children IsNot Nothing AndAlso oClass.Children.Count > 0 Then
                GetPgcCtas(oClass, oCtas)
            End If
        Next
    End Sub

    Public Function YearMonthsCount() As Integer
        Return DTOYearMonth.MonthsDiff(_yearMonthFrom, _yearMonthTo)
    End Function


    Public Function getKpi(id As DTOKpi.Ids) As DTOKpi
        Dim retval As New DTOKpi
        retval.Id = id
        retval.Caption = id.ToString.Replace("_", " ")
        Select Case id
            Case DTOKpi.Ids.Activo_Corriente
                retval.YearMonths = YearMonths(DTOPgcClass.Cods.aAB_Activo_Corriente)
            Case DTOKpi.Ids.Pasivo_Corriente
                retval.YearMonths = YearMonths(DTOPgcClass.Cods.aBC_Pasivo_Corriente)
        End Select
        Return retval
    End Function

    Public Function ExcelBook(oKpis As List(Of DTOKpi), oLang As DTOLang) As ExcelHelper.Book
        Dim retval As New ExcelHelper.Book
        If oKpis.Count > 0 Then retval.Sheets.Add(ExcelSheetKpis(oKpis, oLang))
        retval.Sheets.Add(ExcelSheet(oLang))
        Return retval
    End Function

    Public Function ExcelSheet(oLang As DTOLang) As ExcelHelper.Sheet
        Dim retval As New ExcelHelper.Sheet
        retval.AddColumn("concepte")
        Dim oYearMonths = DTOYearMonth.Range(_yearMonthFrom, _yearMonthTo)
        For Each oYearMonth In oYearMonths
            retval.AddColumn(oYearMonth.Caption(oLang), ExcelHelper.Sheet.NumberFormats.Euro)
        Next

        For Each oClass In _items
            AddRow(retval, oClass, oLang)
        Next
        Return retval
    End Function

    Private Sub AddRow(oSheet As ExcelHelper.Sheet, oClass As DTOPgcClass, oLang As DTOLang)
        Dim oRow = oSheet.AddRow()
        Dim paddingLeft = New String(" ", oClass.GetLevel * 4)
        oRow.AddCell(paddingLeft & oClass.Nom.Tradueix(oLang))
        Dim oYearMonths = DTOYearMonth.Range(_yearMonthFrom, _yearMonthTo)
        If oClass.YearMonths IsNot Nothing AndAlso oClass.YearMonths.Count > 0 Then
            For Each oYearMonth In oYearMonths
                Dim pYearMonth = oClass.YearMonths.FirstOrDefault(Function(x) x.Equals(oYearMonth))
                If pYearMonth Is Nothing Then
                    oRow.AddCell()
                Else
                    oRow.AddCell(pYearMonth.Eur)
                End If
            Next
        End If

        If oClass.Ctas IsNot Nothing AndAlso oClass.Ctas.Count > 0 Then
            paddingLeft = New String(" ", (oClass.GetLevel + 1) * 4)
            For Each oCta In oClass.Ctas
                oRow = oSheet.AddRow()
                oRow.AddCell(String.Format("{0} {1} {2}", paddingLeft, oCta.id, oCta.nom.Tradueix(oLang)))
                For Each oYearMonth In oYearMonths
                    Dim pYearMonth = oCta.yearMonths.FirstOrDefault(Function(x) x.Equals(oYearMonth))
                    If pYearMonth Is Nothing Then
                        oRow.AddCell()
                    Else
                        oRow.AddCell(pYearMonth.Eur)
                    End If
                Next
            Next
        End If
        If oClass.Children IsNot Nothing AndAlso oClass.Children.Count > 0 Then
            For Each oChild In oClass.Children
                AddRow(oSheet, oChild, oLang)
            Next
        End If
    End Sub




    Public Function ExcelSheetKpis(oKpis As List(Of DTOKpi), oLang As DTOLang) As ExcelHelper.Sheet
        Dim retval As New ExcelHelper.Sheet
        retval.AddColumn("any")
        retval.AddColumn("mes")
        For Each oKpi In oKpis
            retval.AddColumn(oKpi.Caption, ExcelHelper.Sheet.NumberFormats.Euro)
        Next
        For i As Integer = 0 To YearMonthsCount() - 1
            Dim oYearMonth = _yearMonthFrom.AddMonths(i)
            Dim oRow = retval.AddRow
            oRow.AddCell(oYearMonth.year)
            oRow.AddCell(oLang.MesAbr(oYearMonth.month))
            For Each oKpi In oKpis
                Dim kYearMonth = oKpi.YearMonths.FirstOrDefault(Function(x) x.Equals(oYearMonth))
                If kYearMonth Is Nothing Then
                    oRow.AddCell()
                Else
                    oRow.AddCell(kYearMonth.Eur)
                End If
            Next
        Next
        Return retval
    End Function
End Class

File: matiasmasso/Mat.NET/DTO/DTOBalance.cs
using MatHelperStd;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOBalance
    {
        public DTOEmp emp { get; set; }
        public DTOYearMonth yearMonthFrom { get; set; }
        public DTOYearMonth yearMonthTo { get; set; }
        public List<DTOPgcClass> items { get; set; }

        public static DTOBalance Factory(DTOEmp oEmp, DTOYearMonth YearMonthFrom, DTOYearMonth YearMonthTo)
        {
            DTOBalance retval = new DTOBalance();
            {
                var withBlock = retval;
                withBlock.emp = oEmp;
                withBlock.yearMonthFrom = YearMonthFrom;
                withBlock.yearMonthTo = YearMonthTo;
                withBlock.items = new List<DTOPgcClass>();
            }
            return retval;
        }

        public List<DTOYearMonth> YearMonths(DTOPgcClass oPgcClass)
        {
            List<DTOPgcCta> oCtas = new List<DTOPgcCta>();
            if (oPgcClass.Ctas != null && oPgcClass.Ctas.Count > 0)
                oCtas = oPgcClass.Ctas;
            else
                GetPgcCtas(oPgcClass, oCtas);
            var oYearMonths = oCtas.SelectMany(x => x.YearMonths).ToList();
            var retval = oYearMonths.GroupBy(g => new { g.Year, g.Month }).Select(group => new DTOYearMonth() { Year = group.Key.Year, Month = group.Key.Month, Eur = group.Sum(x => x.Eur) }).ToList();
            return retval;
        }

        public List<DTOYearMonth> YearMonths(DTOPgcClass.Cods oCod)
        {
            var oCtas = Ctas(oCod);
            var oYearMonths = oCtas.SelectMany(x => x.YearMonths).ToList();
            var retval = oYearMonths.GroupBy(g => new { g.Year, g.Month }).Select(group => new DTOYearMonth() { Year = group.Key.Year, Month = group.Key.Month, Eur = group.Sum(x => x.Eur) }).ToList();
            return retval;
        }

        public List<DTOPgcCta> Ctas(DTOPgcClass.Cods oCod)
        {
            List<DTOPgcCta> retval = new List<DTOPgcCta>();
            var oClass = GetPgcClass(oCod);
            if (oClass != null)
                GetPgcCtas(oClass, retval);
            return retval;
        }

        public DTOPgcClass GetPgcClass(DTOPgcClass.Cods oCod, List<DTOPgcClass> oParentsFrom = null)
        {
            DTOPgcClass retval = null/* TODO Change to default(_) if this is not a reference type */;

            if (oParentsFrom == null)
                oParentsFrom = items;

            foreach (DTOPgcClass oClass in oParentsFrom)
            {
                if (oClass.Cod == oCod)
                {
                    retval = oClass;
                    break;
                }
                else if (oClass.Children.Count > 0)
                {
                    retval = GetPgcClass(oCod, oClass.Children);
                    if (retval != null)
                        break;
                }
            }
            return retval;
        }

        public void GetPgcCtas(DTOPgcClass oParentFrom, List<DTOPgcCta> oCtas)
        {
            foreach (DTOPgcClass oClass in oParentFrom.Children)
            {
                if (oClass.Ctas != null && oClass.Ctas.Count > 0)
                    oCtas.AddRange(oClass.Ctas);
                if (oClass.Children != null && oClass.Children.Count > 0)
                    GetPgcCtas(oClass, oCtas);
            }
        }

        public int YearMonthsCount()
        {
            return DTOYearMonth.MonthsDiff(yearMonthFrom, yearMonthTo);
        }


        public DTOKpi getKpi(DTOKpi.Ids id)
        {
            DTOKpi retval = new DTOKpi();
            retval.Id = id;
            retval.Caption = id.ToString().Replace("_", " ");
            switch (id)
            {
                case DTOKpi.Ids.Activo_Corriente:
                    {
                        retval.YearMonths = YearMonths(DTOPgcClass.Cods.aAB_Activo_Corriente);
                        break;
                    }

                case DTOKpi.Ids.Pasivo_Corriente:
                    {
                        retval.YearMonths = YearMonths(DTOPgcClass.Cods.aBC_Pasivo_Corriente);
                        break;
                    }
            }
            return retval;
        }

        public MatHelper.Excel.Book ExcelBook(List<DTOKpi> oKpis, DTOLang oLang)
        {
            MatHelper.Excel.Book retval = new MatHelper.Excel.Book();
            if (oKpis.Count > 0)
                retval.Sheets.Add(ExcelSheetKpis(oKpis, oLang));
            retval.Sheets.Add(ExcelSheet(oLang));
            return retval;
        }

        public MatHelper.Excel.Sheet ExcelSheet(DTOLang oLang)
        {
            MatHelper.Excel.Sheet retval = new MatHelper.Excel.Sheet();
            retval.AddColumn("concepte");
            var oYearMonths = DTOYearMonth.Range(yearMonthFrom, yearMonthTo);
            foreach (var oYearMonth in oYearMonths)
                retval.AddColumn(oYearMonth.Caption(oLang), MatHelper.Excel.Cell.NumberFormats.Euro);

            foreach (var oClass in items)
                AddRow(retval, oClass, oLang);
            return retval;
        }

        private void AddRow(MatHelper.Excel.Sheet oSheet, DTOPgcClass oClass, DTOLang oLang)
        {
            var oRow = oSheet.AddRow();
            var paddingLeft = new string(' ', oClass.GetLevel() * 4);
            oRow.AddCell(paddingLeft + oClass.Nom.Tradueix(oLang));
            var oYearMonths = DTOYearMonth.Range(yearMonthFrom, yearMonthTo);
            if (oClass.YearMonths != null && oClass.YearMonths.Count > 0)
            {
                foreach (var oYearMonth in oYearMonths)
                {
                    var pYearMonth = oClass.YearMonths.FirstOrDefault(x => x.Equals(oYearMonth));
                    if (pYearMonth == null)
                        oRow.AddCell();
                    else
                        oRow.AddCell(pYearMonth.Eur);
                }
            }

            if (oClass.Ctas != null && oClass.Ctas.Count > 0)
            {
                paddingLeft = new string(' ', (oClass.GetLevel() + 1) * 4);
                foreach (var oCta in oClass.Ctas)
                {
                    oRow = oSheet.AddRow();
                    oRow.AddCell(string.Format("{0} {1} {2}", paddingLeft, oCta.Id, oCta.Nom.Tradueix(oLang)));
                    foreach (var oYearMonth in oYearMonths)
                    {
                        var pYearMonth = oCta.YearMonths.FirstOrDefault(x => x.Equals(oYearMonth));
                        if (pYearMonth == null)
                            oRow.AddCell();
                        else
                            oRow.AddCell(pYearMonth.Eur);
                    }
                }
            }
            if (oClass.Children != null && oClass.Children.Count > 0)
            {
                foreach (var oChild in oClass.Children)
                    AddRow(oSheet, oChild, oLang);
            }
        }




        public MatHelper.Excel.Sheet ExcelSheetKpis(List<DTOKpi> oKpis, DTOLang oLang)
        {
            MatHelper.Excel.Sheet retval = new MatHelper.Excel.Sheet();
            retval.AddColumn("any");
            retval.AddColumn("mes");
            foreach (var oKpi in oKpis)
                retval.AddColumn(oKpi.Caption, MatHelper.Excel.Cell.NumberFormats.Euro);
            for (int i = 0; i <= YearMonthsCount() - 1; i++)
            {
                var oYearMonth = yearMonthFrom.Addmonths(i);
                var oRow = retval.AddRow();
                oRow.AddCell(oYearMonth.Year);
                oRow.AddCell(oLang.MesAbr((int)oYearMonth.Month));
                foreach (var oKpi in oKpis)
                {
                    var kYearMonth = oKpi.YearMonths.FirstOrDefault(x => x.Equals(oYearMonth));
                    if (kYearMonth == null)
                        oRow.AddCell();
                    else
                        oRow.AddCell(kYearMonth.Eur);
                }
            }
            return retval;
        }
    }
}
